{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2ad14cb3",
   "metadata": {},
   "source": [
    "# Data Importation and Representation\n",
    "\n",
    "**Authors:**\n",
    "- Tyrone Lee, Department of Biomedical Informatics, Harvard Medical School (tyrone_lee@hms.harvard.edu)\n",
    "- Tram Nguyen, Department of Biomedical Informatics, Harvard Medical School (Tram_Nguyen@hms.harvard.edu)\n",
    "- Pascal Notin, Department of Systems Biology, Harvard Medical School\n",
    "- Aaron W Kollasch, Department of Systems Biology, Harvard Medical School\n",
    "- Debora Marks, Department of Systems Biology, Harvard Medical School\n",
    "- Ludwig Geistlinger, Department of Biomedical Informatics, Harvard Medical School\n",
    "\n",
    "**Package:** ProteinGymPy\n",
    "\n",
    "**Date:** November 17, 2025"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93695b19",
   "metadata": {},
   "source": [
    "## Setup\n",
    "\n",
    "Load the module and all required dependencies used in the vignette."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f09f013",
   "metadata": {},
   "outputs": [],
   "source": [
    "import proteingympy as pg\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Set display options for better visualization\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.width', None)\n",
    "\n",
    "# Configure matplotlib\n",
    "%matplotlib inline\n",
    "plt.rcParams['figure.figsize'] = (12, 8)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44e351fa",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "Predicting the effects of mutations in proteins is critical to many\n",
    "applications, from understanding genetic disease to designing novel proteins to\n",
    "address our most pressing challenges in climate, agriculture and healthcare.\n",
    "Despite an increase in machine learning-based protein modeling methods,\n",
    "assessing the effectiveness of these models is problematic due to the use of\n",
    "distinct, often contrived, experimental datasets and variable performance across\n",
    "different protein families.\n",
    "\n",
    "[ProteinGym](https://proteingym.org/) is a large-scale and holistic set of\n",
    "benchmarks specifically designed for protein fitness prediction and design\n",
    "curated by @Notin2023.\n",
    "It encompasses both a broad collection of over 250 standardized deep mutational\n",
    "scanning (DMS) assays, spanning millions of mutated sequences, as well as\n",
    "curated clinical datasets providing high-quality expert annotations about\n",
    "mutation effects. Furthermore, ProteinGym reports the performance of a diverse\n",
    "set of over 70 high-performing models from various subfields (eg., mutation\n",
    "effects, inverse folding) into a unified benchmark.\n",
    "\n",
    "ProteinGym datasets are publicly available as a community resource both on\n",
    "[Zenodo](https://doi.org/10.5281/zenodo.13932632) and the official [ProteinGym\n",
    "website](https://proteingym.org/) under the MIT license.\n",
    "\n",
    "### Available datasets\n",
    "\n",
    "The `ProteinGymPy` package provides the following analysis-ready datasets from\n",
    "ProteinGym:\n",
    "\n",
    "1.  DMS assay scores from 217 assays measuring the impact of all possible amino\n",
    "    acid substitutions across 186 proteins. The dataset can be obtained using\n",
    "    the `get_dms_substitution_data()` function\n",
    "\n",
    "2.  AlphaMissense pathogenicity scores for \\~1.6 M substitutions in the\n",
    "    ProteinGym DMS data. The data is provided with `get_alphamissense_proteingym_data()`.\n",
    "\n",
    "3.  Reference file containing metadata associated with the 217 DMS assays, such\n",
    "    as taxon, protein sequence length, UniProt ID, etc. with `get_dms_metadata()`.\n",
    "\n",
    "4.  Five model performance metrics (\"AUC\", \"MCC\", \"NDCG\", \"Spearman\",\n",
    "    \"Top_recall\") for 79 models across 217 assays calculated on DMS\n",
    "    substitutions in a zero-shot setting. The data can be obtained with\n",
    "    `get_zero_shot_benchmark_data()`.\n",
    "\n",
    "5.  Model scores on the DMS substitutions for 79 models in the zero-shot\n",
    "    setting. Load with `get_zero_shot_substitution_data()`.\n",
    "\n",
    "6.  Two model performance metrics (\"Spearman\", and \"MSE\") for 12 models across\n",
    "    217 assays calculated on DMS substitutions in a semi-supervised setting. Load in this data with `get_supervised_metrics()`.\n",
    "\n",
    "7.  Model scores on the DMS substitutions for 12 semi-supervised models with 3\n",
    "    folding schemes: contiguous, modulo, and random. Loaded in with\n",
    "    `get_supervised_substitution_data()` and by changing the \"fold_scheme\" argument,\n",
    "    respectively.\n",
    "\n",
    "8.  PDB files for 197 protein structures, to be used in the `plot_structure()`\n",
    "    3D visualization function.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4686e1d",
   "metadata": {},
   "source": [
    "\n",
    "## Data import\n",
    "\n",
    "[ProteinGym data on Zenodo](https://zenodo.org/records/15293562) can be obtained through ProteinGymPy via the API.\n",
    "\n",
    "### DMS data\n",
    "\n",
    "Deep mutational scanning is an experimental technique that provides\n",
    "experimental data on the fitness effects of all possible single mutations in\n",
    "a protein (Fowler et al. 2024).\n",
    "\n",
    "For each position in a protein, the amino acid residue is mutated and the\n",
    "fitness effects are recorded. While most mutations tend to be deleterious, some\n",
    "can enhance protein activity. In addition to analyzing single mutations, this\n",
    "method can also examine the effects of multiple mutations, yielding insights\n",
    "into protein structure and function. Overall, DMS scores provide a detailed map\n",
    "of how changes in a protein's sequence affect its function, offering valuable\n",
    "\n",
    "Datasets in `ProteinGymPy` can be easily loaded with built-in functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ccfa60c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the DICTIONARY of DMS scores for 217 assays\n",
    "dms_data = pg.get_dms_substitution_data()\n",
    "print(type(dms_data)) # equivalent to R list\n",
    "print(dms_data.keys()) # keys are the assay names -> elements names of the R list\n",
    "print(len(dms_data)) # 217 DMS assays"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb1741fb",
   "metadata": {},
   "source": [
    "View an example of one DMS assay."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38485121",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Grab the first assay\n",
    "first_key = list(dms_data.keys())[0]\n",
    "print(type(dms_data[first_key])) # Now we see this is a PANDAS DataFrame\n",
    "print(dms_data[first_key].head()) # Display the first few rows of the DataFrame\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dadb6b6f",
   "metadata": {},
   "source": [
    "For each DMS assay, the columns show the UniProt protein identifier, the DMS\n",
    "experiment assay identifier, the amino acid substitution at a given sequence \n",
    "position, the mutated protein sequence, the recorded DMS score, and a binary \n",
    "DMS score bin categorizing whether the mutation has an effect on fitness \n",
    "(1) or not (0). For details, see the `get_dms_substitution_data` documentation or reference \n",
    "publication from Notin et al. 2023.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ccb7064",
   "metadata": {},
   "outputs": [],
   "source": [
    "help(pg.get_dms_substitution_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1f0df46",
   "metadata": {},
   "source": [
    "## Model benchmarking\n",
    "\n",
    "The function `pg.benchmark_models()` can be used to compare\n",
    "performance across several variant effect prediction models when using the \n",
    "DMS data as ground truth. This function takes in one of the five\n",
    "available metrics, and and compares the performance of up to 5 out of the \n",
    "79 available models.\n",
    "\n",
    "In the zero-shot setting, the effects of mutations on fitness are predicted \n",
    "without relying on ground-truth labels for the protein of interest. \n",
    "Robust zero-shot performance is particularly informative when labels are\n",
    "subject to several biases or scarcely available (e.g., labels for rare genetic\n",
    "pathologies).\n",
    "\n",
    "Model performance was evaluated across 5 metrics:\n",
    "\n",
    "1.  Spearman's rank correlation coefficient (default metric)\n",
    "2.  Area Under the ROC Curve (AUC)\n",
    "3.  Matthews Correlation Coefficient (MCC), most suitable for bimodal DMS \n",
    "measurements\n",
    "4.  Normalized Discounted Cumulative Gains (NDCG)\n",
    "5.  Top K Recall (top 10% of DMS values)\n",
    "\n",
    "To account for the fact that certain protein functions are overrepresented \n",
    "in the list of proteins assayed with DMS (e.g., thermostability), these metrics \n",
    "were first calculated within groups of proteins with similar functions. The \n",
    "final value of the metric is then the average of these averages, giving each \n",
    "functional group equal weight. The final values are referred to as the \n",
    "'corrected average'.\n",
    "\n",
    "Due to the often non-linear relationship between protein function and organism\n",
    "fitness, the Spearmanâ€™s rank correlation coefficient is typically an appropriate choice for \n",
    "evaluating model performance against experimental measurements. However, in \n",
    "situations where DMS measurements exhibit a bimodal profile, rank correlations \n",
    "may not be the optimal choice. Therefore, additional metrics are also provided, \n",
    "such as the Area Under the ROC Curve (AUC) and the Matthews Correlation \n",
    "Coefficient (MCC), which compare binarized model scores and experimental \n",
    "measurements. Furthermore, for certain goals (e.g., optimizing functional \n",
    "properties of designed proteins), it is more important that a model is able to \n",
    "correctly identify the most functional protein variants, rather than properly \n",
    "capture the overall distribution of all assayed variants. For such scenarios, \n",
    "it is beneficial to use the Normalized Discounted Cumulative Gains (NDCG) which \n",
    "prioritizes models that return high scores for sequences with high DMS value \n",
    "(corresponding to strong gain in fitness). Alternatively, the Top K Recall \n",
    "(with K being set to the top 10% of DMS values) can also be informative for such \n",
    "scenarios.\n",
    "\n",
    "To view all available zero-shot models, use the function: `pg.available_models()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d309a936",
   "metadata": {},
   "outputs": [],
   "source": [
    "zmodels = pg.available_models()\n",
    "print(zmodels)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b84610b7",
   "metadata": {},
   "source": [
    "Now, plot the AUC metric for 5 models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59dd0297",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = pg.benchmark_models(metric = \"AUC\", \n",
    "    models = [\"GEMME\", \"CARP_600K\", \"EVmutation\", \"VESPA\", \"ProtGPT2\"])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94a52e76",
   "metadata": {},
   "source": [
    "Here, GEMME performed the best, achieving highest AUC of the 5 selected models. \n",
    "If not specified by the user, Spearman correlation is used as the default \n",
    "metric. For more information about the models and metrics, see the\n",
    "function documentation for `benchmark_models()`.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d071a84",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "- Notin, P., et al. (2023). ProteinGym: Large-Scale Benchmarks for Protein Design and Fitness Prediction. *Advances in Neural Information Processing Systems*, 36.\n",
    "- Fowler, D. M. and Fields, S. (2014). Deep mutational scanning: a new style of protein science. *Nature Methods*, 11, 801--807.\n",
    "- Chan, K. K., et al. (2020). Engineering human ACE2 to optimize binding to the spike protein of SARS coronavirus 2. *Science*, 369(6508), 1261-1265.\n",
    "- Lee, J. M., et al. (2018). Deep mutational scanning of hemagglutinin helps predict evolutionary fates of human H3N2 influenza variants. *Proceedings of the National Academy of Sciences*, 115(35), E8276-E8285.\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ProteinGymPy (3.13.5)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
